### ISSUE TITLE
Move the IsolateNameServer from dart:ui to dart:isolate

### ISSUE THREAD

**Issue by marcelgarus** [dart-lang/sdk#44495]
The [`IsolateNameServer`](https://api.flutter.dev/flutter/dart-ui/IsolateNameServer-class.html) is currently in dart:ui, which means it only can be used on the Flutter runtime. I'm working on a non-Flutter project where it would come in handy.

Is there a particular reason it's in dart:ui, although it would intuitively better fit in dart:isolate?

**Comment by lrhn** [comment]
The `dart:ui` library is maintained by the Flutter team, and the `dart:isolate` library is owned by the Dart team.

As a Dart library maintiner, I'd probably *not* put that functionality into `dart:isolate`. 
It's utility functionality, not something most Dart isolate users will necessarily need (something which might be different for Flutter users if the framework uses it).
If the Dart team was to own the functionality, I'd probably put it into `package:isolate` instead. This keeps the core platform libraries slimmer and the functionality easier to maintain (it's *so* much easier to make changes to versioned packages than to the SDK libraries). We'd probably not add it, and instead just recommend using the isolate [registry](https://pub.dev/documentation/isolate/latest/isolate.registry/RegistryManager-class.html) functionality instead. It's more complicated than the `IsolateNameServer`, but if you store one `SendPort` value per `String` tag, it should support the same functionality.

**Comment by marcelgarus** [comment]
Oh, I see. That dart:ui and dart:isolate are owned by different teams explains a lot (for example, I also wondered why `hashValues` was in dart:ui).

To give a bit more context: My use case is a [pure-Dart database](https://github.com/marcelgarus/chest) where I would need a global namespace: If the developer opens a database in one isolate and also opens the same database in another isolate, I'd like both of those calls to connect to and communicate with the same backend isolate, which would handle accessing the file and exchanging changes.

So if I understand correctly, that's currently not possible in pure Dart and I would have to depend on Flutter?

**Comment by lrhn** [comment]
Correct. It's not possible in pure Dart for an arbitrary isolate to know about another isolate without there being some pre-existing communication. That would generally mean that spawning a new isolate would require sending at least one send port along with the spawn request, so that it's possible to set up communications. If you just do `Isolate.spwan(someFunction, null)` then there is no available way to start communicating with that isolate using pure isolate communication.

You could set up an HTTP server on a known port and communicate using that, but then you won't have any send ports, and will need to serialize all the data yourself.

**Comment by marcelgarus** [comment]
Okay, that's what I thought. The HTTP approach sounds interesting, I might look into that.

Would it theoretically be possible that in the future the `IsolateNameServer` gets migrated to dart:isolate or is that completely off the table?
Given that the functionality exists, it would semantically make more sense for it to be in dart:isolate.
And conceptually, it also makes sense that this functionality exists at all: Having global state (like the global table from names to ports) sounds like a bad idea for most everyday use-cases, but in the end, a Dart program is one entity in the OS that somehow needs to access OS resources – be that files, ports etc. Managing this access using functional-style pure isolates doesn't work (because files, ports etc. inherently have an identity), so there should be a way to manage that access in Dart. Working around this by having to rely on the OS for this functionality (like using ports or files to communicate with other isolates) indicates that something's missing in the DartVM.

**Comment by lrhn** [comment]
There is no *technical* issue with moving the class and implementation into the Dart SDK from the Flutter SDK.
Whether we want it ... that's a design decision. It sort-of breaks the *isolation* of isolates, but so does file I/O.
I can see that pragmatically it works. It's a sort of "registry" for send ports, which is all you need to set up communication between any two isolates, so it's a minimal feature which solves the problem.

I probably wouldn't *name* it `IsolateNameServer`. It's not serving isolates. More like `PortRegistry`.

**Comment by marcelgarus** [comment]
> It sort-of breaks the isolation of isolates, but so does file I/O.

Exactly. My issue is that there are valid use-cases for this functionality (like managing file access, port access, communicating with other programs, etc). Currently, developers have the following two choices:

* Only make their package available for Flutter, which harms the general Dart ecosystem.
* Work around this by using ports or files for communicating, which is ugly and less efficient than `SendPort`s (serialization/deserialization is necessary).

I completely agree that `PortRegistry` is a better fitting name.

> Whether we want it ... that's a design decision.

What would be the next steps? Is there a design committee or something that looks over this?

**Comment by lrhn** [comment]
Since the isolate library is owned by the VM team, I'll make this an area-vm enhancement issue.

**Comment by marcelgarus** [comment]
Maybe interesting: Elixir also uses an actor approach for concurrency. The equivalent to `Isolate`s are `Process`es and you can register their ids/ports in a global registry: https://hexdocs.pm/elixir/master/Process.html#register/2

**Comment by maks** [comment]
@lrhn I thought I could avoid asking for the VM team to have to spend their developer resources on this by
 implementing this for myself using FFI based on the [code already there in the Flutter teams implementation](https://github.com/flutter/engine/tree/main/lib/ui/isolate_name_server).

But sadly what I quickly discovered is that while there is a way to get a [native port](https://api.dart.dev/stable/3.0.5/dart-ffi/NativePort.html) intended for use by C code, there is no way to then make use of this on the Dart side to get back a SendPort. 

So I think the only way this can be done is by moving it into the Dart SDK due to needing to use the Dart native extensions unless there is some way to serialise and then deserialise Dart objects (specifically SendPorts) for use with FFI that I'm missing?

I think I remember seeing somewhere that the FFI could eventually replace the older native extensions mechanism, so this is I guess also an example where that currently isn't possible.

PS. My use is pretty much as per @MarcelGarus comment, of recreating a similar functionality of Erlang/Elixir but I would prefer without needing to pass around a Registry to every spawned Isolate as required by the [pure Dart Registry in the Isolate package](https://pub.dev/documentation/isolate/latest/isolate.registry/Registry-class.html).
